src/animation/KeyframeTrack.js

```diff
@@ -106,79 +182,109 @@ THREE.KeyframeTrack.prototype = {
 
 	// removes keyframes before and after animation without changing any values within the range [startTime, endTime].
 	// IMPORTANT: We do not shift around keys to the start of the track time, because for interpolated keys this will change their values
- 	trim: function( startTime, endTime ) {
+	trim: function( startTime, endTime ) {
+
+		var times = this.times;
+		var nKeys = times.length;
 
 		var firstKeysToRemove = 0;
-		for ( var i = 1; i < this.keys.length; i ++ ) {
-			if ( this.keys[i] <= startTime ) {
-				firstKeysToRemove ++;
-			}
+		for ( var i = 1; i !== nKeys; ++ i ) {
+
+			if ( times[i] <= startTime ) ++ firstKeysToRemove;
+
 		}
 
 		var lastKeysToRemove = 0;
-		for ( var i = this.keys.length - 2; i > 0; i ++ ) {
-			if ( this.keys[i] >= endTime ) {
-				lastKeysToRemove ++;
-			} else {
-				break;
-			}
+		for ( var i = nKeys - 2; i !== 0; -- i ) {
+
+			if ( times[i] >= endTime ) ++ lastKeysToRemove;
+			else break;
+
 		}
 
 		// remove last keys first because it doesn't affect the position of the first keys (the otherway around doesn't work as easily)
-		if ( ( firstKeysToRemove + lastKeysToRemove ) > 0 ) {
-			this.keys = this.keys.splice( firstKeysToRemove, this.keys.length - lastKeysToRemove - firstKeysToRemove );;
-		}
+		if( ( firstKeysToRemove + lastKeysToRemove ) !== 0 ) {
 
-		return this;
+			var from = firstKeysToRemove;
+			var to = nKeys - lastKeysToRemove - firstKeysToRemove;
 
-	},
+			this.times = THREE.AnimationUtils.arraySlice( times, from, to );
 
-	/* NOTE: This is commented out because we really shouldn't have to handle unsorted key lists
-	         Tracks with out of order keys should be considered to be invalid.  - bhouston
-	sort: function() {
+			var values = this.values;
+			var stride = this.getValueSize();
+			this.values = THREE.AnimationUtils.arraySlice( values, from * stride, to * stride );
 
-		this.keys.sort( THREE.KeyframeTrack.keyComparer );
+		}
 
 		return this;
 
-	},*/
+	},
 
 	// ensure we do not get a GarbageInGarbageOut situation, make sure tracks are at least minimally viable
 	// One could eventually ensure that all key.values in a track are all of the same type (otherwise interpolation makes no sense.)
 	validate: function() {
 
-		var prevKey = null;
+		if ( this.getValueSize() % 1 !== 0 ) {
```

Comment:
> Could you add fract method to THREE.Math and add a unit test or something?

*Please*, let's *not* do that!

I dislike convenience mumbo-jumbo that can be expressed directly in terms of the programming language: It takes already way too long for the JIT to get an app to full framerate. Fine-grained factorization, as appealing as it sometimes seems, makes this situation worse. Same goes for GLSL, BTW: I know examples where using many functions causes insane compilation / optimization times with some widely-used implementations - easily over a minute (!).

Could we just learn the relevant parts of the languages and use them directly, instead of trying to shape them into something foolproof? :-)

The behavior of the modulus operator is clearly defined for JavaScript. Since there is no `fmod` in `Math`, no JS implementation can have gotten around it. However, **I rewrote the offending condition to `valueSize - Math.floor( valueSize ) !== 0`**.

> The other way to be more obvious, as well as faster, would be ( this.values.length % this.times.length ) === 0.

The assumption that values and times actually correspond was intended to be a default, not a requirement! **I also gave `.validate` a proper cleanup**, as it was not consistently following that scheme. `Interpolant` also only touches values, in methods that can be overridden separately, from `getAt` BTW.